sky/engine/core/script/dart_controller.cc

```diff
@@ -68,6 +68,16 @@ DartController::~DartController() {
   if (dom_dart_state_) {
     // Don't use a DartIsolateScope here since we never exit the isolate.
     Dart_EnterIsolate(dom_dart_state_->isolate());
+
+    // Run the framework's shutdown handler.
+    {
+      DartApiScope dart_api_scope;
+      Dart_Handle ui_library = Dart_LookupLibrary(ToDart("dart:ui"));
+      if (!LogIfError(ui_library)) {
+        DartInvokeField(ui_library, "_onShutdown", {});
+      }
+    }
```

Comment:
> I'm a little curious what we intend to use this callback for.

It's to close mojo handles that we hold for the lifetime of the isolate.

> In the general case we might not be able to promise that callbacks run before shutdown occurs (say if the OS kills our process).

Presumably that will also result in all open Mojo handles being closed.

> Is the Dart isolate holding onto any heavyweight state that needs to be freed sooner than it might otherwise?

It's hard to know what it's holding onto.  Depends what's on the other end of the pipes.